hydroflow_datalog_core/src/lib.rs

```diff
@@ -238,11 +214,137 @@ fn generate_rule(
         Pipeline::Link(PipelineLink {
             lhs: Box::new(parse_quote!(#out_name #out_indexing)), // out_name[idx]
             arrow: parse_quote!(->),
-            rhs: Box::new(after_join),
+            rhs: Box::new(after_join_and_send),
         }),
     ));
 }
 
+fn apply_aggregations(
+    rule: &Rule,
+    out_expanded: &IntermediateJoinNode,
+    diagnostics: &mut Vec<Diagnostic>,
+) -> Pipeline {
+    let mut aggregations = vec![];
+    let mut group_by_exprs = vec![];
+    let mut agg_exprs = vec![];
+
+    for field in rule.target.fields.iter() {
+        let expr: syn::Expr = if let Some(col) = out_expanded
+            .variable_mapping
+            .get(&syn::Ident::new(&field.ident().name, Span::call_site()))
+        {
+            let source_col_idx = syn::Index::from(*col);
+            parse_quote!(row.#source_col_idx)
+        } else {
+            diagnostics.push(Diagnostic::spanned(
+                Span::call_site(),
+                Level::Error,
+                format!(
+                    "Could not find column {} in RHS of rule",
+                    &field.ident().name
+                ),
+            ));
+            parse_quote!(())
+        };
+
+        match field {
+            TargetExpr::Ident(_) => {
+                group_by_exprs.push(expr);
+            }
+            TargetExpr::Aggregation(_) => {
+                aggregations.push(field);
+                agg_exprs.push(expr);
+            }
+        }
+    }
+
+    let flattened_tuple_type = &out_expanded.tuple_type;
+
+    if agg_exprs.is_empty() {
+        parse_quote!(map(|row: #flattened_tuple_type| (#(#group_by_exprs, )*)))
+    } else {
+        let agg_initial_values = agg_exprs
+            .iter()
+            .map(|_| parse_quote!(None))
+            .collect::<Vec<syn::Expr>>();
+        let agg_initial: syn::Expr = parse_quote!((#(#agg_initial_values, )*));
+
+        let group_by_input_types = group_by_exprs
+            .iter()
+            .map(|_| parse_quote!(_))
+            .collect::<Vec<syn::Type>>();
+        let group_by_input_type: syn::Type = parse_quote!((#(#group_by_input_types, )*));
+
+        let agg_input_types = agg_exprs
+            .iter()
+            .map(|_| parse_quote!(_))
+            .collect::<Vec<syn::Type>>();
+        let agg_input_type: syn::Type = parse_quote!((#(#agg_input_types, )*));
+
+        let agg_types = agg_exprs
+            .iter()
+            .map(|_| parse_quote!(Option<_>))
+            .collect::<Vec<syn::Type>>();
+        let agg_type: syn::Type = parse_quote!((#(#agg_types, )*));
```

Comment:
Probably ought to make a `repeat_tuple` helper or something, `repeat_tuple(quote!(None), agg_exprs.len())`